src/log.rs

```diff
@@ -0,0 +1,396 @@
+        user_data: glib_sys::gpointer,
+    ) {
+        let log_domain: GString = from_glib_borrow(log_domain);
+        let log_level = from_glib(log_level);
+        let message: GString = from_glib_borrow(message);
+        let callback: &P = &*(user_data as *mut _);
+        (*callback)(log_domain.as_str(), &log_level, message.as_str());
+    }
+    let log_func = Some(log_func_func::<P> as _);
+    unsafe extern "C" fn destroy_func<P: Fn(&str, &LogLevelFlags, &str) + Send + Sync + 'static>(
+        data: glib_sys::gpointer,
+    ) {
+        let _callback: Box_<P> = Box_::from_raw(data as *mut _);
+    }
+    let destroy_call4 = Some(destroy_func::<P> as _);
+    let super_callback0: Box_<P> = log_func_data;
+    unsafe {
+        glib_sys::g_log_set_handler_full(
+            log_domain.to_glib_none().0,
+            log_levels.to_glib(),
+            log_func,
+            Box_::into_raw(super_callback0) as *mut _,
+            destroy_call4,
+        )
+    }
+}
+
+pub fn log_set_handler<P: Fn(&str, &LogLevelFlags, &str) + Send + Sync + 'static>(
+    log_domain: Option<&str>,
+    log_levels: LogLevelFlags,
+    log_func: P,
+) -> u32 {
+    let log_func_data: Box_<P> = Box_::new(log_func);
+    unsafe extern "C" fn log_func_func<
+        P: Fn(&str, &LogLevelFlags, &str) + Send + Sync + 'static,
+    >(
+        log_domain: *const libc::c_char,
+        log_level: glib_sys::GLogLevelFlags,
+        message: *const libc::c_char,
+        user_data: glib_sys::gpointer,
+    ) {
+        let log_domain: GString = from_glib_borrow(log_domain);
+        let log_level = from_glib(log_level);
+        let message: GString = from_glib_borrow(message);
+        let callback: &P = &*(user_data as *mut _);
+        (*callback)(log_domain.as_str(), &log_level, message.as_str());
+    }
+    let log_func = Some(log_func_func::<P> as _);
+    let super_callback0: Box_<P> = log_func_data;
+    unsafe {
+        glib_sys::g_log_set_handler(
+            log_domain.to_glib_none().0,
+            log_levels.to_glib(),
+            log_func,
+            Box_::into_raw(super_callback0) as *mut _,
+        )
+    }
+}
+
+static PRINT_HANDLER: Lazy<Mutex<Option<Box_<Box_<dyn Fn(GString) + Send + Sync + 'static>>>>> =
+    Lazy::new(|| Mutex::new(None));
+
+pub fn set_print_handler<P: Fn(GString) + Send + Sync + 'static>(func: Option<P>) {
+    unsafe extern "C" fn func_func(string: *const libc::c_char) {
+        match PRINT_HANDLER.lock() {
+            Ok(handler) => {
+                if let Some(ref handler) = *handler {
+                    let string: GString = from_glib_borrow(string);
+                    (*handler)(string)
+                } else {
+                    panic!("PRINT_HANDLER cannot be None!");
+                }
+            }
+            Err(_) => {
+                // should we log something here?
+            }
+        }
+    }
+    let callback = if func.is_some() {
+        Some(func_func as _)
+    } else {
+        None
```

Comment:
Split this into a `set_print_handler()` and `unset_print_handler()`. Otherwise this is annoying to use for callers

Reply:
I'm not sure to follow the issue here?

Reply:
```rust
Some(|| {

})
```

is annoying to type/read, and we can easily make it nicer

Reply:
I'm still completely lost... What's the link with `(un)set_print_handler`?

Reply:
Currently it's `set_print_handler(Some(|...| { ...})` and that sucks. It should be `set_print_handler(|..| { ... })` and `unset_print_handler()`.

Reply:
Oh. I see! Makes sense. You weren't talking about this part of the code specifically but of the function more globally!

Reply:
Yes, sorry :)